### ISSUE TITLE
Improve exception logging for API Gateway Local Service

### ISSUE THREAD

**Issue by lfpetrini** [awslabs/aws-sam-cli#1947]
This is a very simple request, I didn't know if it could be considered a bug since it's just related to logging, but I have spent hours trying to figure out what is wrong with the response I'm sending back from my lambda function and the only error message I get is this one:

> [32mREPORT RequestId: ac7ae76f-bdb4-1dd8-b184-6a7d2ad94f20     Init Duration: 544.64 ms        Duration: 23.88 ms      Billed Duration: 100 ms Memory Size: 128 MB     Max Memory Used: 56 MB      [0m
> **<class 'samcli.local.apigw.local_apigw_service.LambdaResponseParseException'>**

I'm new to SAM so I decided to take a look at the [source code](https://github.com/awslabs/aws-sam-cli/blob/develop/samcli/local/apigw/local_apigw_service.py).
Even though the exceptions are raised with very informative messages, at the end it will just log its type instead of anything useful (line 214):
```
except LambdaResponseParseException:
            LOG.error(str(LambdaResponseParseException))
            return ServiceErrorResponses.lambda_failure_response()
```

I suppose an \__str__ function should be added to the exception, and it should log its instance instead of its type.

**Comment by awood45** [comment]
That makes sense, we'd be happy to take a PR for this as well.

**Comment by mndeveci** [comment]
Hi @lfpetrini ,

Thanks for raising this issue up. I took a look at the concern here, and try to re-produce it by throwing an exception inside the lambda function. But I can see the exception has been logged into console.

Example code piece that I added inside handler function to throw the exception;
```
if true {
	panic(errors.New("Error message here"))
}
```
And the logs here;

```
Fetching lambci/lambda:go1.x Docker container image......
Mounting /Volumes/unix/workplace/other/sam-internal-go-app/.aws-sam/build/HelloWorldFunction as /var/task:ro,delegated inside runtime container
START RequestId: 34908c02-e2e5-13be-8dfc-fe1422c06da1 Version: $LATEST
END RequestId: 34908c02-e2e5-13be-8dfc-fe1422c06da1
REPORT RequestId: 34908c02-e2e5-13be-8dfc-fe1422c06da1  Init Duration: 266.24 ms        Duration: 471.41 ms     Billed Duration: 500 ms    Memory Size: 128 MB     Max Memory Used: 26 MB
Lambda returned empty body!
Invalid API Gateway Response Keys: {'errorType', 'errorMessage', 'stackTrace'} in {'errorType': 'errorString', 'errorMessage': 'Error message here', 'stackTrace': ['github.com/aws/aws-lambda-go@v1.13.3/lambda/function.go:35 (*Function).Invoke.func1', 'runtime/panic.go:679 gopanic', 'hello-world/main.go:44 handler', 'reflect/value.go:460 Value.call', 'reflect/value.go:321 Value.Call', 'github.com/aws/aws-lambda-go@v1.13.3/lambda/handler.go:124 NewHandler.func1', 'github.com/aws/aws-lambda-go@v1.13.3/lambda/handler.go:24 lambdaHandler.Invoke', 'github.com/aws/aws-lambda-go@v1.13.3/lambda/function.go:67 (*Function).Invoke', 'reflect/value.go:460 Value.call', 'reflect/value.go:321 Value.Call', 'net/rpc/server.go:377 (*service).call', 'runtime/asm_amd64.s:1357 goexit']}
<class 'samcli.local.apigw.local_apigw_service.LambdaResponseParseException'>
```

Can you elaborate your issue here? Which version of sam-cli you are using, and what was the response that is returned by the lambda function?

**Comment by lfpetrini** [comment]
Hi there @mndeveci , thanks for taking a look into the issue.  
The version is 0.47.0. I've managed to get it working by changing the parent handler (I'm using java, by the way). But the issue is actually about UX. If you notice in your output you get<br/>
`<class 'samcli.local.apigw.local_apigw_service.LambdaResponseParseException'>`<br/>
If you look at the [source code](https://github.com/awslabs/aws-sam-cli/blob/develop/samcli/local/apigw/local_apigw_service.py), those exceptions are created with descriptions about the root cause, or the underlying exceptions. In most cases, the message is also logged just before the exception is raised and logged, so that would be fine. But in two places (currently lines 271 and 274) the message is not logged and it's just passed as parameter to the constructor of LambdaResponseParseException
```python
try:
    json_output = json.loads(lambda_output)
except ValueError:
    raise LambdaResponseParseException(f"Lambda response must be valid json: {lambda_output}")

if not isinstance(json_output, dict):
    raise LambdaResponseParseException(f"Lambda returned {type(json_output)} instead of dict")
```

So when they are caught, it is only logging the type (and not the instance or the message about the issue):
```python
LOG.error(str(LambdaResponseParseException))
```

In summary, instead of logging the exception like:<br/>
`<class 'samcli.local.apigw.local_apigw_service.LambdaResponseParseException'>`<br/>
I believe the desired behaviour would be to also include the message associated with the exception, something like:<br/>
`samcli.local.apigw.local_apigw_service.LambdaResponseParseException: Lambda returned <class 'int'> instead of dict`

**Comment by mndeveci** [comment]
Thanks for detailed explanation.

I saw that we are throwing `LambdaResponseParseException` from multiple places and in some of them, we log the original exception but for some of them it is not been logged. And as you pointed out in the [source code](https://github.com/awslabs/aws-sam-cli/blob/develop/samcli/local/apigw/local_apigw_service.py#L217-L219), we are only logging exception type, not the actual exception message nor stack trace.

Moving exception logging into single place will fix this issue.

**Comment by mndeveci** [comment]
Patched with [v1.0.0](https://github.com/awslabs/aws-sam-cli/releases/tag/v1.0.0)
